zookeeper-server/src/main/java/org/apache/zookeeper/server/watch/IWatchManager.java

```diff
@@ -34,6 +34,22 @@
      */
     boolean addWatch(String path, Watcher watcher);
 
+    /**
+     * Add watch to specific path.
+     *
+     * @param path znode path
+     * @param watcher watcher object reference
+     * @param watcherMode the watcher mode to use
+     *
+     * @return true if the watcher added is not already present
+     */
+    default boolean addWatch(String path, Watcher watcher, WatcherMode watcherMode) {
+        if (watcherMode == WatcherMode.DEFAULT_WATCHER_MODE) {
+            return addWatch(path, watcher);
+        }
+        throw new UnsupportedOperationException();  // custom implementations must defeat this
```

Comment:
This path will be followed with OptimizedWatchManager.
How will it bubble up to the client?

I see two ways:
- implement the method everywhere (it will have a cost!) 
- add a test case about the behaviour of a client that it is trying to use this feature with a server that does not support it because it has not the default WatchManager impl

Thoughts?

Reply:
@eolivelli test added. See: https://github.com/Randgalt/zookeeper/blob/ZOOKEEPER-1416/zookeeper-server/src/test/java/org/apache/zookeeper/test/UnsupportedAddWatcherTest.java